DRILL-7230: Add README.md with instructions for release and release scripts (#1937)
apache/drill: docs/dev/Release.md

# Drill Release Process

1. ## Setup release final cut-off date

    1. Start mail thread on Apache dev list with the subject: 
        ```

>        [DISCUSS] Drill 1.17.0 release

Reviewer: Not that this is an example

        ```

        Body example:
        ```
        Hi Drillers,
         
        It's been several months since the last release and it is time to do the next one. I am volunteering to be the release manager.
         
        If there are any issues on which work is in progress, that you feel we *must* include in the release, please post in reply to this thread. Based on your input we'll define release cut off date.
         
        Kind regards
         
        Thanks
        ```
    2. Gather list of Jiras that need to be included in the release, discuss possible blockers.
    3. Identify final cut-off date.

>     4. Before starting the release send a reminder to dev list that merge in Apache Drill master is forbidden during

Reviewer: ```suggestion
    4. Before starting the release send a reminder to dev list that merge into Apache Drill master is forbidden during
```

        3. Generate GPG key using the **Apache email address** if it wasn't done before:
            1. Run `gpg --gen-key` and follow the instructions.
            2. Remember your passphrase.
            3. Save your key pair:
                ```
               gpg --output mygpgkey_pub.gpg --armor --export MY_KEY_ID
               gpg --output mygpgkey_sec.gpg --armor --export-secret-key MY_KEY_ID
               ```
        4. Or import existing GPG key pair if required:
            ```
            gpg --import mygpgkey_pub.gpg
            gpg --allow-secret-key-import --import mygpgkey_sec.gpg
            ```
        5. Have another committer signed your key (add to the trust chain).
            Apache advises to do it at 
            [key signing parties](https://www.apache.org/dev/release-signing.html#key-signing-party).
        6. Make sure default key is the key generated using the Apache email.
        7. Publish your public key to a public server (e.g. http://pgp.surfnet.nl or http://pgp.mit.edu)
            using Web-UIs or CLI (for example using the following command:
            `gpg --keyserver hkp://pool.sks-keyservers.net --trust-model always --send-keys`)

>         8. Add your key signature to the `KEYS` file in the Drill sources repo.

Reviewer: Indicate that KEYS are in the root.

            Instruction on how to create the key signature is embedded at the beginning of the KEYS file.

>         9. Add your key signature to the `KEYS` file in `https://dist.apache.org` (PMC permissions required for this):

Reviewer: ```suggestion
        9. Add your key signature to the `KEYS` file into the `https://dist.apache.org` (PMC permissions required for this):
```

            - `svn co https://dist.apache.org/repos/dist/release/drill ~/src/release/drill-dist`
            - Update `KEYS` file with required changes
            - Commit changes: `svn ci -m "Add FirstName LastName's key to KEYS file"`
    2. ### Setting up LDAP credentials
        1. Go to http://id.apache.org and log in using your Apache login and password.
        2. Scroll to the bottom to the 'SSH Key' entry and add the SSH key from your `~/.ssh/id_rsa.pub` file.
            If you don't have the SSH key, you should generate it by running `ssh-keygen`.
        3. Note that you can add more than 1 SSH key corresponding to multiple machines.
        4. Enter your Apache password and submit changes.

>         5. Verify that you can do an sftp to the apache server by running the following: `sftp <username>@home.apache.org`

Reviewer: ```suggestion
        5. Verify that you can do an sftp to the Apache server by running the following: `sftp <username>@home.apache.org`
```

>     3. ### SVN

Reviewer: `9. Add your key signature to the `KEYS` already requires SVN looks like having this step after point 9 is strange
How to setup up SVN access?

        You also need access to Apache SVN. (You need to be a PMC member for this).
    4. ### Setup Maven
        1. Apache's Maven repository access is documented here:
            http://www.apache.org/dev/publishing-maven-artifacts.html
            http://www.apache.org/dev/publishing-maven-artifacts.html#dev-env.
        2. For deploying maven artifacts, you must add entries to your `~/.m2/settings.xml` file as described here:
            http://www.apache.org/dev/publishing-maven-artifacts.html#dev-env.
        3. In order to encrypt the LDAP password, see here:
            http://maven.apache.org/guides/mini/guide-encryption.html.
            The encrypted master password should be added to `~/.m2/security-settings.xml`.
        4. Add encrypted password to `settings.xml` and `security-settings.xml`:
            1. run `mvn --encrypt-master-password` and encrypted master password to `security-settings.xml` file;
            2. run `mvn --encrypt-password` and encrypted Apache LDAP password to `settings.xml` file.
    5. Check that `NOTICE` file in sources has the current copyright year.

>     6. Make sure you are using JDK 8 (not 9 or 13).

Reviewer: ```suggestion
    6. Make sure you are using JDK 8.
```

3. ## Manual Release process
    1. Setup GPG Password env variable:
        ```
        read -s GPG_PASSPHRASE
        ```
    2. Tell GPG how to read a password from your terminal:
        ```
        export GPG_TTY=$(tty)
        ```

>     3. Clean up old release directory and clone a new drill repository

Reviewer: ```suggestion
    3. Clean up old release directory and clone a new Drill repository
```

        ```
        cd ~/src/release && \
        rm -rf drill/ && \
        git clone git@github.com:apache/drill.git && \
        cd drill
        ```
    4. Checkout the current release:
        ```
        git checkout <git commit>
        ```

>     5. Make sure build with tests succeed along with license checks:

Reviewer: ```suggestion
    5. Make sure build with tests succeeded along with license checks:
```

        ```
        mvn install -Drat.skip=false -Dlicense.skip=false
        ```
    6. Enforce memory options:
        ```
        MAVEN_OPTS='-Xmx4g -XX:MaxPermSize=512m'
        ```

>     7. Clear any release history

Reviewer: ```suggestion
    7. Clear any release history:
```

        ```
        mvn release:clean -Papache-release -DpushChanges=false -DskipTests
        ```

>     8. Delete outgoing release tag if it exists

Reviewer: ```suggestion
    8. Delete outgoing release tag if it exists:
```

        ```
        git push --delete origin drill-1.17.0
        git tag -d drill-1.17.0
        ```

>     9. Do the release preparation

Reviewer: It's just in some points you have `:` in some don't.
Please be consistent.

        ```
    10. Make sure to change Drill version to the proper one 
    11. Check that `target` folder contains the following files (with the correct version number):
        ```
        apache-drill-1.17.0-src.tar.gz 
        apache-drill-1.17.0-src.tar.gz.asc
        apache-drill-1.17.0-src.tar.gz.sha512
        apache-drill-1.17.0-src.zip
        apache-drill-1.17.0-src.zip.asc 
        apache-drill-1.17.0-src.zip.sha512
        ```
    12. Verify signature, ensure that GPG key for Apache was used (see details at
        https://www.apache.org/info/verification.html#CheckingSignatures) and checksum
        (see details at https://www.apache.org/info/verification.html#CheckingHashes):
         ```
        ./tools/release-scripts/checksum.sh target/apache-drill-1.17.0-src.tar.gz
        ./tools/release-scripts/checksum.sh target/apache-drill-1.17.0-src.zip
        ./tools/release-scripts/checksum.sh distribution/target/apache-drill-1.17.0.tar.gz
        ```
    13. Push the new tag to your personal github repository

>         You should see 2 new commits on top of the branch and a new `drill-1.17.0` tag

Reviewer: Please add example

        ```
        git push origin drill-1.17.0
        ```

>     14. Perform the release but do the commits to personal repo

Reviewer: ```suggestion
    14. Perform the release but do the commits into personal repo
```

        (Since Apache doesn't allow rollbacks on master if mistakes are made)
        ```
        mvn release:perform -DconnectionUrl=scm:git:git@github.com:vvysotskyi/drill.git -DskipTests -Darguments="-Dgpg.passphrase=${GPG_PASSPHRASE} -DskipTests=true -DconnectionUrl=scm:git:git@github.com:vvysotskyi/drill.git"
        ```

>         If you want additionally check resulting archives and jars, add `-Dmaven.deploy.skip=true` flag to avoid deploying jars to the Nexus repository:

Reviewer: ```suggestion
        If you want to additionally check resulting archives and jars, add `-Dmaven.deploy.skip=true` flag to avoid deploying jars to the Nexus repository:
```

        mvn deploy -Papache-release -DskipTests -Dgpg.passphrase=${GPG_PASSPHRASE}
        ```
    16. Copy release files to a local release staging directory
        ```
        cp ~/src/release/drill/target/target/checkout/apache-drill-1.17.0-src.tar.gz* ~/release/1.17.0-rc0/ && \ 
        cp ~/src/release/drill/target/target/checkout/apache-drill-1.17.0.zip* ~/release/1.17.0-rc0/ \ 
        cp ~/src/release/drill/target/checkout/distribution/target/apache-drill-1.17.0.tar.gz* ~/release/1.17.0-rc0/ \ 
        ```
    17. Check if the artifacts are signed properly:
        ```
        ./tools/release-scripts/checksum.sh ~/release/1.17.0-rc0/apache-drill-1.17.0-src.tar.gz
        ./tools/release-scripts/checksum.sh ~/release/1.17.0-rc0/apache-drill-1.17.0-src.zip
        ./tools/release-scripts/checksum.sh ~/release/1.17.0-rc0/apache-drill-1.17.0.tar.gz
        ```
    18. Copy release files to a directory on `home.apache.org` for voting:
        ```
        scp ~/release/1.17.0-rc0/* <username>@home.apache.org:~/public_html/drill/releases/1.17.0/rc0
        ```

4. ## Automated release
    Run the following script:
    ```
    tools/release-scripts/release.sh
    ```

>     The release script will push the maven artifacts to the maven staging repo.

Reviewer: ```suggestion
    The release script will push the maven artifacts to the Maven staging repo.
```

5. ## Publish release candidate and vote
    1. Go to the [Apache Maven staging repo](https://repository.apache.org/) and close the new jar release
        
        This step is done in the Maven GUI.
    2. Start vote (vote should last at least 72 hours).

>         Email should be sent to Drill user and dev mailing lists.

Reviewer: ```suggestion
        Email should be sent to the Drill user and dev mailing lists.
```

            
            Thanks to all who voted. A new release will be tagged as 1.17.0 - RC1 and
            will be available for voting soon.
            ```
        2. Go to the [Apache Maven staging repo](https://repository.apache.org/) and drop the jar release.
        3. Delete the release tag locally and remotely:
            ```
            git push --delete origin drill-1.17.0
            git tag -d drill-1.17.0
            ```
    4. If the vote passes, finish the release:
        1. Count votes (email to the dev mailing list):

            Subject:
            ```
            [RESULT] [VOTE] Release Apache Drill 1.12.0 RC1
            ```
            Body:
            ```
            The vote passes. Thanks to everyone who has tested the release candidate and given their comments and votes. Final tally:
            
            3x +1 (binding): Arina, Aman, Parth
            
            5x +1 (non-binding): Vitalii, Holger, Prasad, Vova, Charles 
            
            No 0s or -1s.
            
            I'll start process for pushing the release artifacts and send an announcement once propagated.
            
            Kind regards
            ```

>         2. Add the release to [dist.apache.org](https://dist.apache.org/repos/dist/release/drill/) and delete the old version, keeping two most recent.

Reviewer: ```suggestion
        2. Add the release to the [dist.apache.org](https://dist.apache.org/repos/dist/release/drill/) and delete the old version, keeping two most recent.
```

            This can only done by a PMC member:
            ```
            svn co https://dist.apache.org/repos/dist/release/drill ~/src/release/drill-dist
            cd ~/src/release/drill-dist
            cp -r ~/release/1.17.0-rc0 1.17.0
            svn add 1.XX.0
            svn commit --message "Upload Apache Drill 1.17.0 release."
            svn delete 1.15-1.0
            ```

>         3. Go to [Apache Maven staging repo](https://repository.apache.org/) and promote release to production.

Reviewer: ```suggestion
        3. Go to the [Apache Maven staging repo](https://repository.apache.org/) and promote release to production.
```

>         3. Go to [Apache Maven staging repo](https://repository.apache.org/) and promote release to production.

Reviewer: ```suggestion
        3. Go to [Apache Maven staging repo](https://repository.apache.org/) and promote release to the production.
```

>         4. Create branch and tag for this release and update drill version in master (if used automated scripts, tag will be like this `drill-1.11.0`. Branch should be named as `1.11.0`).

Reviewer: ```suggestion
        4. Create branch and tag for this release and update Drill version in master (if used automated scripts, tag will be like this `drill-1.11.0`. Branch should be named as `1.11.0`).
```

        5. Wait 24 hours, check download mirrors.
        6. Post release
            1. "What's New" for the new release
            2. Update Apache JIRA and add release date for this release. Add new release tag if not already there.

>             3. Update Drill Web site:

Reviewer: Add point to post on Twitter :)

Author: Great point! Thanks, added.

                    1. Checkout to the release version: `git checkout drill-1.17.0`
                    2. Run execute the following command to generate Javadocs in `target/site/apidocs` directory:
                        ```
                        mvn install javadoc:aggregate -DskipTests
                        ```
                    3. Checkout to the `gh-pages` branch where sources for site are placed:
                        ```
                        git checkout gh-pages
                        ```
                    4. Remove JavaDocs for previous version:
                        ```
                        rm -rf apidocs
                        ```
                    5. Copy new JavaDocs and commit changes:
                        ```
                        mv target/site/apidocs/* .
                        ```
                6. Instructions how to build and deploy Web site may be found here:
                    https://github.com/apache/drill/blob/gh-pages/README.md
            4. A PMC member needs to update the release date for new release here:
                        https://reporter.apache.org/addrelease.html?drill

>             5. Send announcement to `dev@drill.apache.org` and `announce@apache.org` mailing lists

Reviewer: ```suggestion
            5. Send the announcement to the `dev@drill.apache.org` and `announce@apache.org` mailing lists
```

    4. ### Setup Maven
        1. Apache's Maven repository access is documented here:
            http://www.apache.org/dev/publishing-maven-artifacts.html
            http://www.apache.org/dev/publishing-maven-artifacts.html#dev-env.
        2. For deploying Maven artifacts, you must add entries to your `~/.m2/settings.xml` file as described here:
            http://www.apache.org/dev/publishing-maven-artifacts.html#dev-env.
        3. In order to encrypt the LDAP password, see here:
            http://maven.apache.org/guides/mini/guide-encryption.html.
            The encrypted master password should be added to `~/.m2/security-settings.xml`.
        4. Add encrypted password to `settings.xml` and `security-settings.xml`:
            1. run `mvn --encrypt-master-password` and add an encrypted master password to `security-settings.xml` file;
            2. run `mvn --encrypt-password` and add an encrypted Apache LDAP password to `settings.xml` file.
    5. Check that `NOTICE` file in sources has the current copyright year.
    6. Make sure you are using JDK 8.

> 3. ## Manual Release process (don't worry, after this section described an automated release process)

Reviewer: ```suggestion
3. ## Manual Release process (this section is more for information how release process is 
performed, release manager should use `automated release process` described later). 
```

    15. Deploy the release commit:
        ```
        git checkout drill-1.17.0
        mvn deploy -Papache-release -DskipTests -Dgpg.passphrase=${GPG_PASSPHRASE}
        ```
    16. Copy release files to a local release staging directory:
        ```
        cp ~/src/release/drill/target/target/checkout/apache-drill-1.17.0-src.tar.gz* ~/release/1.17.0-rc0/ && \ 
        cp ~/src/release/drill/target/target/checkout/apache-drill-1.17.0.zip* ~/release/1.17.0-rc0/ \ 
        cp ~/src/release/drill/target/checkout/distribution/target/apache-drill-1.17.0.tar.gz* ~/release/1.17.0-rc0/ \ 
        ```
    17. Check if the artifacts are signed properly:
        ```
        ./tools/release-scripts/checksum.sh ~/release/1.17.0-rc0/apache-drill-1.17.0-src.tar.gz
        ./tools/release-scripts/checksum.sh ~/release/1.17.0-rc0/apache-drill-1.17.0-src.zip
        ./tools/release-scripts/checksum.sh ~/release/1.17.0-rc0/apache-drill-1.17.0.tar.gz
        ```
    18. Copy release files to a directory on `home.apache.org` for voting:
        ```
        scp ~/release/1.17.0-rc0/* <username>@home.apache.org:~/public_html/drill/releases/1.17.0/rc0
        ```

> 4. ## Automated release

Reviewer: ```suggestion
4. ## Automated release process
```

    Run the following script:
    ```
    tools/release-scripts/release.sh
    ```
    The release script will push the maven artifacts to the Maven staging repo.
5. ## Publish release candidate and vote
    1. Go to the [Apache Maven staging repo](https://repository.apache.org/) and close the new jar release.

>         This step is done in the Maven GUI.

Reviewer: Maybe add info link: https://central.sonatype.org/pages/releasing-the-deployment.html#locate-and-examine-your-staging-repository

        4. Create branch and tag for this release and update Drill version in master (if used automated scripts, tag will be like this `drill-1.11.0`. Branch should be named as `1.11.0`).
            Add release description in GitHub [Drill releases page](https://github.com/apache/drill/releases):
            open release tag, press `Edit tag` and add Release title like `Apache Drill 1.17.0` and specify release description.
        5. After new tag is pushed, check that DockerHub triggered job to build and publish new image for the release to https://hub.docker.com/repository/docker/apache/drill.
        5. Wait 24 hours, check download mirrors.
        6. Post release:
            1. "What's New" for the new release.
            2. Update Apache JIRA and add release date for this release. Add a new release tag if not already there.
            3. Update Drill Web site:
                1. Generate release notes for Drill: https://confluence.atlassian.com/jira/creating-release-notes-185729647.html
                    and create a MarkDown file for the release notes - post to the site the day of the release.

>                 2. Write the blog post and push it out to the Apache Drill website so it can be referenced in the announcement.

Reviewer: Maybe include commit examples for point 2 and 3

                3. Update the Download page to state the current release, like 1.6 and change the download link for the
                    Apache mirror site and direct file downloads.
                4. Update the data in `_data/version.json` - making sure to either match your release notes url to the one
                    below (`docs/apache-drill-1-6-0-release-notes`) or update it with the different URL:
                    ```
                    {
                      "display_version": "1.6",
                      "full_version": "1.6.0",
                      "release_date": "February 16, 2016",
                      "blog_post":"/blog/2016/02/16/drill-1.6-released",
                      "release_notes": "https://drill.apache.org/docs/apache-drill-1-6-0-release-notes/"
                    }
                    ```

>                 5. Build and publish JavaDocs to site:

Reviewer: ```suggestion
                5. Build and publish JavaDocs to the Drill site:
```
